mgmtd/mgmt_be_adapter.c

```diff
@@ -135,178 +114,91 @@ mgmt_be_find_adapter_by_name(const char *name)
 }
 
 static void mgmt_register_client_xpath(enum mgmt_be_client_id id,
-				       const char *xpath, uint subscribed)
+				       const char *xpath, bool config)
 {
-	struct mgmt_be_xpath_map *map;
+	struct mgmt_be_xpath_map **maps, *map;
+
+	maps = config ? &be_cfg_xpath_map : &be_oper_xpath_map;
 
-	darr_foreach_p (mgmt_xpath_map, map)
+	darr_foreach_p (*maps, map) {
 		if (!strcmp(xpath, map->xpath_regexp)) {
-			map->subscr_info[id] = subscribed;
+			map->clients = (1u << id);
 			return;
 		}
+	}
 	/* we didn't find a matching entry */
-	map = darr_append(mgmt_xpath_map);
+	map = darr_append(*maps);
 	map->xpath_regexp = XSTRDUP(MTYPE_MGMTD_XPATH, xpath);
-	map->subscr_info[id] = subscribed;
+	map->clients = (1ul << id);
 }
 
 /*
- * Load the initial mapping from static init map
+ * initial the combined maps from per client maps
  */
 static void mgmt_be_xpath_map_init(void)
 {
-	struct mgmt_be_client_xpath *init, *end;
 	enum mgmt_be_client_id id;
+	const char *const *init;
 
 	MGMTD_BE_ADAPTER_DBG("Init XPath Maps");
 
 	FOREACH_MGMTD_BE_CLIENT_ID (id) {
-		init = mgmt_client_xpaths[id].xpaths;
-		end = init + mgmt_client_xpaths[id].nxpaths;
-		for (; init < end; init++) {
-			MGMTD_BE_ADAPTER_DBG(" - XPATH: '%s'", init->xpath);
-			mgmt_register_client_xpath(id, init->xpath,
-						   init->subscribed);
+		/* Initialize the common config init map */
+		for (init = be_client_xpaths[id]; init && *init; init++) {
+			MGMTD_BE_ADAPTER_DBG(" - CFG XPATH: '%s'", *init);
+			mgmt_register_client_xpath(id, *init, true);
+		}
+
+		/* Initialize the common oper init map */
+		for (init = be_client_oper_xpaths[id]; init && *init; init++) {
+			MGMTD_BE_ADAPTER_DBG(" - OPER XPATH: '%s'", *init);
+			mgmt_register_client_xpath(id, *init, false);
 		}
 	}
 
-	MGMTD_BE_ADAPTER_DBG("Total XPath Maps: %u", darr_len(mgmt_xpath_map));
+	MGMTD_BE_ADAPTER_DBG("Total Cfg XPath Maps: %u",
+			     darr_len(be_cfg_xpath_map));
+	MGMTD_BE_ADAPTER_DBG("Total Oper XPath Maps: %u",
+			     darr_len(be_oper_xpath_map));
 }
 
 static void mgmt_be_xpath_map_cleanup(void)
 {
 	struct mgmt_be_xpath_map *map;
 
-	darr_foreach_p (mgmt_xpath_map, map)
+	darr_foreach_p (be_cfg_xpath_map, map)
 		XFREE(MTYPE_MGMTD_XPATH, map->xpath_regexp);
-	darr_free(mgmt_xpath_map);
-}
-
-static int mgmt_be_eval_regexp_match(const char *xpath_regexp,
-				     const char *xpath)
-{
-	int match_len = 0, re_indx = 0, xp_indx = 0;
-	int rexp_len, xpath_len;
-	bool match = true, re_wild = false, xp_wild = false;
-	bool delim = false, enter_wild_match = false;
-	char wild_delim = 0;
+	darr_free(be_cfg_xpath_map);
 
-	rexp_len = strlen(xpath_regexp);
-	xpath_len = strlen(xpath);
-
-	/*
-	 * Remove the trailing wildcard from the regexp and Xpath.
-	 */
-	if (rexp_len && xpath_regexp[rexp_len-1] == '*')
-		rexp_len--;
-	if (xpath_len && xpath[xpath_len-1] == '*')
-		xpath_len--;
```

Comment:
@idryzhov here it was stripping trailing `*`s. But if you look over the rest of this code, it's quite broken, the only reason things worked correctly is b/c the broken code mostly return "Yes!" for matches regardless of the actual comparison. The daemons would then just ignore what they didn't understand.